doc/designs/rbcd.md

```diff
@@ -0,0 +1,385 @@
+objectClass `resourceDelegation` is present in the same entry, KDB driver will
+use information from the `memberPrincipal` attribute to check whether a service
+asking for S4U2Proxy extension is allowed to send a forwarded user ticket to
+this service.
+
+This approach allows implementing MS-SFU's `ServicesAllowedToReceiveForwardedTicketsFrom`
+information.
+
+Management of the resource-based constrained delegation is integrated into `ipa
+host` and `ipa service` commands.
+
+|Command                         |Description                                                 |
+|--------------------------------|------------------------------------------------------------|
+|service-add-delegation          |Add new resource delegation to a service                    |
+|service-allow-add-delegation    |Allow users, groups, hosts or host groups to handle a resource delegation of this service.|
+|service-disallow-add-delegation |Disallow users, groups, hosts or host groups to handle a resource delegation of this service.|
+|service-remove-delegation       |Remove resource delegation from a service|
+|host-add-delegation             |Add new resource delegation to a host|
+|host-allow-add-delegation       |Allow users, groups, hosts or host groups to handle a resource delegation of this host.|
+|host-disallow-add-delegation    |Disallow users, groups, hosts or host groups to handle a resource delegation of this host.|
+|host-remove-delegation          |Remove resource delegation from a host|
+
+### Implementation
+
+#### IPA API commands
+
+In both general constrained delegation and resource-based constrained
+delegation rules, targets, and delegation permission details are expressed with
+`memberPrincipal` attribute. Since IPA's standard `LDAPAddMember` and
+`LDAPRemoveMember` classes operate on DNs for the members, they need to be
+overridden to represent Kerberos principals. The principal might be not present
+in the IPA realm and thus cannot be represented as an LDAP object to which DN
+could be constructed.
+
+##### General constrained delegation
+
+In case of general constrained delegation a special handling is added to
+`LDAPAddMember` and `LDAPRemoveMember` classes to handle `memberPrincipal`.
+The add/remove methods assume, and require via asserts, that all members be a
+DN. `get_member_dns()` needs to ignore any `memberPrincipal` values and return
+only the DN-based values when adding targets to rules to let the standard
+mechanics of `LDAP*Member` do their work.
+
+In order to handle the `memberPrinicpal` values a `post_callback()` is required.
+This also means that there be at worst two writes per membership update.
+Given that this feature is not expected to be frequently used then speed and
+efficiency are not a factor.
+
+Similarly, we enforce that only a target is a member of a rule, and not
+another rule. That would be an undefined relationship. To do this each member
+needs to be retrieved and evaluated before adding as a member.
```

Comment:
Does this mean a target may only be a member of a single rule?

Reply:
No, the target can be a member of multiple rules. The default configuration re-uses the same rule for several targets. The text is borrowed from the original S4U2Proxy design document and is really meant to say that `ipaAllowedTarget` values are DNs pointing to the delegation targets and not for anything else.

```
dn: cn=ipa-http-delegation,cn=s4u2proxy,cn=etc,$SUFFIX
changetype: add
objectClass: ipaKrb5DelegationACL
objectClass: groupOfPrincipals
objectClass: top
cn: ipa-http-delegation
memberPrincipal: HTTP/$HOST@$REALM
ipaAllowedTarget: cn=ipa-ldap-delegation-targets,cn=s4u2proxy,cn=etc,$SUFFIX
ipaAllowedTarget: cn=ipa-cifs-delegation-targets,cn=s4u2proxy,cn=etc,$SUFFIX

dn: cn=ipa-ldap-delegation-targets,cn=s4u2proxy,cn=etc,$SUFFIX
changetype: add
objectClass: groupOfPrincipals
objectClass: top
cn: ipa-ldap-delegation-targets
memberPrincipal: ldap/$HOST@$REALM

dn: cn=ipa-cifs-delegation-targets,cn=s4u2proxy,cn=etc,$SUFFIX
changetype: add
objectClass: groupOfPrincipals
objectClass: top
cn: ipa-cifs-delegation-targets

```